File: client/bindings/nodejs/examples/02_generate_addresses.ts
// Copyright 2021-2022 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {
    Client,
    CoinType,
    initLogger,
    SHIMMER_TESTNET_BECH32_HRP,
} from '@iota/client';
require('dotenv').config({ path: '../.env' });

// Run with command:
// node ./dist/02_generate_addresses.js

// In this example we will create addresses from a mnemonic defined in .env
async function run() {
    initLogger();
    if (!process.env.NODE_URL) {
        throw new Error('.env NODE_URL is undefined, see .env.example');
    }

    const client = new Client({
        // Insert your node URL in the .env.
        nodes: [process.env.NODE_URL],
    });

    try {
        if (!process.env.NON_SECURE_USE_OF_DEVELOPMENT_MNEMONIC_1) {
            throw new Error('.env mnemonic is undefined, see .env.example');
        }
        const secretManager = {
            mnemonic: process.env.NON_SECURE_USE_OF_DEVELOPMENT_MNEMONIC_1,
        };

        // Generate public address with custom account index and range.
        const address = await client.generateAddresses(secretManager, {
            accountIndex: 0,
            range: {
                start: 0,
                end: 1,
            },
        });
        console.log('First public address:', address, '\n');

        // Generate an internal address with custom account index and range.
        const internalAddress = await client.generateAddresses(secretManager, {
            accountIndex: 0,
            range: {
                start: 0,
                end: 1,
            },
            internal: true,
        });
        console.log('First internal address:', internalAddress, '\n');

        // Generate addresses with providing all inputs, that way it can also be done offline without a node.
        const offlineGeneratedAddresses = await client.generateAddresses(
            secretManager,
            {
                coinType: CoinType.Shimmer,
                accountIndex: 0,
                range: {
                    start: 0,
                    end: 2,
                },
                internal: false,
                // Generating addresses with client.generateAddresses(secretManager, {}), will by default get the bech32_hrp (Bech32
                // human readable part) from the node info, generating it "offline" requires setting it in the generateAddressesOptions
                bech32Hrp: SHIMMER_TESTNET_BECH32_HRP,
            },
        );
        console.log(
            'List of offline generated public addresses:',
            offlineGeneratedAddresses,
        );
    } catch (error) {
        console.error('Error: ', error);
    }
}

run().then(() => process.exit());

File: client/bindings/python/examples/02_generate_addresses.py
from iota_client import IotaClient, MnemonicSecretManager

# Create an IotaClient instance
client = IotaClient({'nodes': ['https://api.testnet.shimmer.network']})

# In this example we will create addresses from a mnemonic

secret_manager = MnemonicSecretManager("endorse answer radar about source reunion marriage tag sausage weekend frost daring base attack because joke dream slender leisure group reason prepare broken river")

# Generate public address with default account index and range.
addresses = client.generate_addresses(secret_manager, options={})

print('List of generated public addresses:', *addresses, sep='\n')
print()

# Generate public address with custom account index and range.
addresses = client.generate_addresses(secret_manager, {
    "accountIndex": 0,
    "range": {
        "start": 0,
        "end": 4,
    },
})

print('List of generated public addresses:', *addresses, sep='\n')
print()

# Generate internal addresses with custom account index and range.
addresses = client.generate_addresses(secret_manager, {
    "accountIndex": 0,
    "range": {
        "start": 0,
        "end": 4,
    },
    "internal": True,
})

print('List of generated internal addresses:', *addresses, sep='\n')
print()

# Generate addresses with providing all inputs, that way it can also be done offline without a node.
addresses = client.generate_addresses(secret_manager, {
    "coinType": 4219, # Shimmer coin type
    "accountIndex": 0,
    "range": {
        "start": 0,
        "end": 4,
    },
    "internal": False,
    # Generating addresses with client.generateAddresses(secretManager, options={}), will by default get the bech32_hrp (Bech32
    # human readable part) from the node info, generating it "offline" requires setting it in the generateAddressesOptions
    "bech32Hrp": 'rms'
})

print('List of offline generated public addresses:', *addresses, sep='\n')
print()
